secboot/secboot_tpm.go

```diff
@@ -244,3 +264,166 @@ func unlockEncryptedPartition(tpm *sb.TPMConnection, name, device, keyfile, pinf
+			systemdStubParams := sb.SystemdEFIStubProfileParams{
+				PCRAlgorithm:   tpm2.HashAlgorithmSHA256,
+				PCRIndex:       tpmPCR,
+				KernelCmdlines: modelParams.KernelCmdlines,
+			}
+			if err := sbAddSystemdEFIStubProfile(modelProfile, &systemdStubParams); err != nil {
+				return fmt.Errorf("cannot add systemd EFI stub profile: %v", err)
+			}
+		}
+
+		// Add snap model profile
+		if modelParams.Model != nil {
+			snapModelParams := sb.SnapModelProfileParams{
+				PCRAlgorithm: tpm2.HashAlgorithmSHA256,
+				PCRIndex:     tpmPCR,
+				Models:       []*asserts.Model{modelParams.Model},
+			}
+			if err := sbAddSnapModelProfile(modelProfile, &snapModelParams); err != nil {
+				return fmt.Errorf("cannot add snap model profile: %v", err)
+			}
+		}
+
+		modelPCRProfiles = append(modelPCRProfiles, modelProfile)
+	}
+
+	var pcrProfile *sb.PCRProtectionProfile
+	if numModels > 1 {
+		pcrProfile = sb.NewPCRProtectionProfile().AddProfileOR(modelPCRProfiles...)
+	} else {
+		pcrProfile = modelPCRProfiles[0]
+	}
+
+	// Provision the TPM as late as possible
+	if err := tpmProvision(tpm, params.TPMLockoutAuthFile); err != nil {
+		return err
+	}
+
+	// If we are re-installing, there will be an a sealed keyfile that already
+	// exists on the ubuntu-seed partition that we should remove before trying
+	// to seal the new key, otherwise trying to seal the new key will fail
+	if osutil.FileExists(params.KeyFile) {
+		logger.Noticef("keyfile from previous install exists, deleting")
+		if err := os.Remove(params.KeyFile); err != nil {
+			return fmt.Errorf("cannot delete encryption key from previous install: %v", err)
+		}
+	}
+
+	// Seal key to the TPM
+	creationParams := sb.KeyCreationParams{
+		PCRProfile: pcrProfile,
+		PINHandle:  pinHandle,
+	}
+	if err := sbSealKeyToTPM(tpm, key[:], params.KeyFile, params.PolicyUpdateDataFile, &creationParams); err != nil {
+		return fmt.Errorf("cannot seal data: %v", err)
+	}
+
+	return nil
+}
+
+func tpmProvision(tpm *sb.TPMConnection, lockoutAuthFile string) error {
+	// Create and save the lockout authorization file
+	lockoutAuth := make([]byte, 16)
+	// crypto rand is protected against short reads
+	_, err := rand.Read(lockoutAuth)
+	if err != nil {
+		return fmt.Errorf("cannot create lockout authorization: %v", err)
+	}
+	if err := osutil.AtomicWriteFile(lockoutAuthFile, lockoutAuth, 0600, 0); err != nil {
+		return fmt.Errorf("cannot write the lockout authorization file: %v", err)
+	}
+
+	// TODO:UC20: ideally we should ask the firmware to clear the TPM and then reboot
+	//            if the device has previously been provisioned, see
+	//            https://godoc.org/github.com/snapcore/secboot#RequestTPMClearUsingPPI
+	if err := sbProvisionTPM(tpm, sb.ProvisionModeFull, lockoutAuth); err != nil {
+		logger.Noticef("TPM provisioning error: %v", err)
+		return fmt.Errorf("cannot provision TPM: %v", err)
+	}
+	return nil
+}
+
+func buildLoadSequence(filePaths []string) *sb.EFIImageLoadEvent {
```

Comment:
the implementation here doesn't support any kind of tree like you have detailed below, and instead essentially just takes a single list and chains the things in that list together in order. Do we need support for having multiple specific chains? As it seems the code only seems to ever create a single list of shim -> grub -> grub -> kernel, whereas it seems like we also need to support shim -> grub -> kernel too, no?

Also as the implementation of this function is rather naive I think a better name would be something like `chainLoadSequence` and have the doc-comment make it clear that it just chains together a static list, in order and doesn't do any sort of permutation, etc.

Reply:
Will rename and add comments. This is indeed a very simple implementation that creates a sequential list, and is intended to be used while the exact mechanisms needed for reseal are not determined. For sealing purposes having multiple lists or a tree is equivalent (if the possible chains are the same in both topologies).

To have the exact load chains for different supported modes, shim -> recovery grub -> recovery kernel is also needed for recover mode (and would also be a different chain for the same model).

Reply:
Refactored the function in such a way that it takes path sequences and return EFI image load event sequences that can be directly used in the EFI secure boot policy profile parameters. The actual list of files will have to be determined from the model and gadget, but for now it's just hard coded in bootstrap.go.